src/collector/api/v1/handler/json/report.go

```diff
@@ -0,0 +1,187 @@
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/query/storage"
+	"github.com/m3db/m3/src/x/serialize"
+	"github.com/m3db/m3metrics/metric/id"
+	"github.com/m3db/m3x/instrument"
+)
+
+const (
+	// ReportURL is the url for the report json handler
+	ReportURL = handler.RoutePrefixV1 + "/json/report"
+
+	// ReportHTTPMethod is the HTTP method used with this resource.
+	ReportHTTPMethod = http.MethodPost
+
+	counterType = "counter"
+	gaugeType   = "gauge"
+	timerType   = "timer"
+)
+
+var (
+	errEncoderNoBytes = errors.New("tags encoder has no access to bytes")
+)
+
+type reportHandler struct {
+	reporter       reporter.Reporter
+	encoderPool    serialize.TagEncoderPool
+	decoderPool    serialize.TagDecoderPool
+	instrumentOpts instrument.Options
+}
+
+// NewReportHandler returns a new instance of the report handler.
+func NewReportHandler(
+	reporter reporter.Reporter,
+	encoderPool serialize.TagEncoderPool,
+	decoderPool serialize.TagDecoderPool,
+	instrumentOpts instrument.Options,
+) http.Handler {
+	return &reportHandler{
+		reporter:       reporter,
+		encoderPool:    encoderPool,
+		decoderPool:    decoderPool,
+		instrumentOpts: instrumentOpts,
+	}
+}
+
+// reportRequest represents the report request from the caller.
+type reportRequest struct {
+	Metrics []metricValue `json:"metrics"`
+}
+
+// metricValue is a reportable metric value.
+type metricValue struct {
+	Type  string            `json:"type" validate:"nonzero"`
+	Tags  map[string]string `json:"tags" validate:"nonzero"`
+	Value float64           `json:"value" validate:"nonzero"`
+}
+
+// reportResponse represents the report response.
+type reportResponse struct {
+	Reported int `json:"reported"`
+}
+
+func (h *reportHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	req, err := h.parseRequest(r)
+	if err != nil {
+		handler.Error(w, err.Inner(), err.Code())
+		return
+	}
+
+	for _, metric := range req.Metrics {
+		id, err := h.newMetricID(metric)
+		if err != nil {
+			handler.Error(w, err.Inner(), err.Code())
+			return
+		}
+
+		if err := h.reportMetric(id, metric); err != nil {
+			handler.Error(w, err.Inner(), err.Code())
+			return
+		}
+	}
+
+	resp := &reportResponse{Reported: len(req.Metrics)}
+	handler.WriteJSONResponse(w, resp, h.instrumentOpts.ZapLogger())
+}
+
+func (h *reportHandler) parseRequest(r *http.Request) (*reportRequest, *handler.ParseError) {
+	body := r.Body
+	if r.Body == nil {
+		err := fmt.Errorf("empty request body")
+		return nil, handler.NewParseError(err, http.StatusBadRequest)
+	}
+
+	defer body.Close()
+
+	req := new(reportRequest)
+	if err := json.NewDecoder(body).Decode(req); err != nil {
+		return nil, handler.NewParseError(err, http.StatusBadRequest)
+	}
+
+	return req, nil
+}
+
+func (h *reportHandler) newMetricID(metric metricValue) (id.ID, *handler.ParseError) {
+	tags := make(models.Tags, 0, len(metric.Tags))
+	for n, v := range metric.Tags {
+		tags = tags.AddTag(models.Tag{Name: []byte(n), Value: []byte(v)})
```

Comment:
I guess the tag name has to be a string after JSON deserialization because you can't have []byte be a map key, but can the JSON map be `map[string][]byte` so we don't have to convert it here?

Reply:
Nah unfortunately JSON decoder expects `[]byte` types to be base64 encoded strings, so will make this endpoint very hard to use.  

This is a super low volume test endpoint, so don't fret too much about perf here.